capture/src/billing_limits.rs

```diff
@@ -0,0 +1,181 @@
+    ) -> Result<BillingLimiter, RedisError> {
+        let limited = limited.unwrap_or_else(|| HashSet::new());
+        let limited = Arc::new(RwLock::new(limited));
+
+        Ok(BillingLimiter { nodes, limited })
+    }
+    
+    async fn fetch_limited(client: &ClusterClient) -> Result<Vec<String>, RedisError>{
+        let mut conn = client.get_async_connection().await?;
+        let now = time::OffsetDateTime::now_utc().unix_timestamp();
+
+        conn.zrangebyscore(format!("{QUOTA_LIMITER_CACHE_KEY}events"), now, "+Inf").await
+
+    }
+
+    /// Spawn the tokio updater task
+    ///
+    /// Note: If no nodes are set, the billing limiter falls back to limiting a fixed set. No
+    /// updates.
+    ///
+    /// This will return an error if called more than once, globally. The background task will run
+    /// for the duration of the capture.
+    ///
+    /// If for some reason the task fails (eg lost redis connection), we fail open by clearing the
+    /// local cache and terminating the task.
+    pub async fn start_updater(&self) -> Result<(), RedisError> {
+        if self.nodes.is_empty() {
+            return Ok(())
+        }
+
+        let limited = self.limited.clone();
+        let client = ClusterClient::new(self.nodes.clone())?;
+
+        tokio::spawn(async move {
+            // use an interval rather than sleep
+            // this takes into account the last execution time, so it will wait for 5 seconds since
+            // the last execution time, not 5 seconds since we start awaiting. hence mut.
+            let mut interval =
+                tokio::time::interval(tokio::time::Duration::from_secs(UPDATE_INTERVAL_SECS));
+
+            // The goal is for the _entire_ iteration of this loop to take UPDATE_INTERVAL_SECS
+            // If our execution takes less time than that, the interval will pad it out.
+            loop {
+                let _span = tracing::debug_span!("billing limit update tick");
+                interval.tick().await;
+
+                let tokens = Self::fetch_limited(&client).await;
+
+                match tokens {
+                    Ok(tokens) => {
+                        tracing::debug!("{} limited tokens fetched from redis", tokens.len());
+
+                        // RAII guard on the write lock
+                        let mut l = limited.write().await;
+                        
+                        for i in tokens {
+                            l.insert(i);
+                        }
+                    }
+                    Err(e) => {
+                        // Oh no! Clear all limits. Something has gone wrong so let's play it safe.
+                        tracing::error!("error fetching limited set from redis: {}", e);
+
+                        let mut l = limited.write().await;
+                        l.clear();
+                    }
+                }
+            }
+        });
+
+        Ok(())
+    }
+
+    pub async fn is_limited(&self, key: &str) -> bool {
```

Comment:
Although the ideal topology will be to have a dedicated capture pool per data type, we won't have that luxury on local devenv & hobby. Plus we might decide to colocate several "beta" data types in a common pool, to reduce the cost of an MVP.

For that, we'll want a `data_type` enum, that is a parameter of that function. I think it's fine to sync all of `QUOTA_LIMITER_CACHE_KEY` for now for simplicity

Reply:
done 👍  just copied it pretty much as is from the python